tests/scenario/test_machine_charm/test_relation_priority.py

```diff
@@ -0,0 +1,158 @@
+        assert not charm._cos.dashboards
+        assert not charm._cos.logs_alerts
+        assert not charm._cos.metrics_alerts
+        assert not charm._cos.metrics_jobs
+        assert not charm._cos.snap_log_plugs
+
+        assert charm._principal_relation
+        assert charm.principal_unit
+
+    set_run_out(mock_run, 0)
+    trigger("start", State(relations=[Relation("juju-info")]), post_event=post_event)
+
+
+@patch("machine_charm.subprocess.run")
+def test_cos_machine_relation(mock_run):
+    def post_event(charm: machine_charm.GrafanaAgentMachineCharm):
+        assert charm._cos.dashboards
+        assert charm._cos.snap_log_plugs
+        assert not charm._cos.logs_alerts
+        assert not charm._cos.metrics_alerts
+        assert not charm._cos.metrics_jobs
+
+        assert charm._principal_relation.name == "cos-machine"
+        assert charm.principal_unit.name == "remote-cos-machine/0"
+
+    set_run_out(mock_run, 0)
+    data = {
+        "config": json.dumps(
+            {
+                "metrics": {
+                    "scrape_jobs": [],
+                    "alert_rules": {},
+                },
+                "logs": {
+                    "targets": ["foo:bar", "baz:qux"],
+                    "alert_rules": {},
+                },
+                "dashboards": {
+                    "dashboards": [
+                        COSMachineProvider._encode_dashboard_content("very long dashboard")
+                    ],
+                },
+            }
+        )
+    }
+    trigger(
+        "start",
+        State(
+            relations=[
+                Relation("cos-machine", remote_app_name="remote-cos-machine", remote_app_data=data)
+            ]
+        ),
+        post_event=post_event,
+    )
+
+
+@patch("machine_charm.subprocess.run")
+def test_both_relations(mock_run):
+    def post_event(charm: machine_charm.GrafanaAgentMachineCharm):
+        assert charm._cos.dashboards
+        assert charm._cos.snap_log_plugs
+        assert not charm._cos.logs_alerts
+        assert not charm._cos.metrics_alerts
+        assert not charm._cos.metrics_jobs
+
+        # we have both, but principal is grabbed from cos-machine
+        assert charm._principal_relation.name == "cos-machine"
+        assert charm.principal_unit.name == "remote-cos-machine/0"
+
+    set_run_out(mock_run, 0)
+    data = {
+        "config": json.dumps(
+            {
+                "metrics": {
+                    "scrape_jobs": [],
+                    "alert_rules": {},
+                },
+                "logs": {
+                    "targets": ["foo:bar", "baz:qux"],
+                    "alert_rules": {},
+                },
+                "dashboards": {
+                    "dashboards": [
+                        COSMachineProvider._encode_dashboard_content("very long dashboard")
+                    ],
+                },
+            }
+        )
+    }
+    trigger(
+        "start",
```

Comment:
All the tests in the file only trigger start. Does this make the tests brittle?
I.e. these seem borderline whitebox tests.

Reply:
Good point, however, what I'm testing here is not the output state but the charm state, as you can tell by the `post_event` callbacks. And also I'm not interested in why the charm is executed, I just want it to initialize to then verify the return values of some of its properties and methods.

So it is brittle in that if we ever stop listening to 'start', we'll have to use some other event to have the charm execute. But it's not brittle in that we're not in fact checking anything specific to the start event in itself. Any other event will do. 
We could parametrize and run these tests for a bunch other events as well (or all of them), but it felt pointless.

Reply:
You're right in that the box here is much whiter than by other scenario tests I've written, but it felt a bit premature to run assertions on the output config given that it's still fluid

Reply:
Not a priority!